### DISCUSSION TITLE
how to use a super admin when team are enabled

### DISCUSSION THREAD

**Discussion by sohailahmad07** [spatie/laravel-permission#2672] [General]
HI when you enable teams then you lose global roles permission like in my apps teams are optional and we the user login he has a `User` role with some permission but i can not access any of those permissions unless i assign the user to a team similarly   i have a `super admin` role which can handle all of these different teams but is not associated with any teams what about that 
 
 
 so is there a way to handle global roles and permission we user teams are enabled

**Comment by erikn69** [comment]
For global roles just set to NULL the `team_id` on the roles table

**Reply by sohailahmad07** [reply]
yes but i can not assign it to any user like i have a role super admin but and team_id is null but i can not assign to any user 

$user->assignRole('super-admin');

i get error team_id can not be null

**Reply by erikn69** [reply]
Do you read docs?

[basic-usage/teams-permissions#content-working-with-teams-permissions](https://spatie.be/docs/laravel-permission/v6/basic-usage/teams-permissions#content-working-with-teams-permissions)

need to add the team to the package before any change, usually at login

just to test do this
```php
setPermissionsTeamId($YOUR_TEAM_ID ?? 1);
$user->assignRole('super-admin');
```

**Reply by sohailahmad07** [reply]
yes I know that work but the super admin does not belong to any team he actually manages all these other teams think of my super admin as an owner or an admin for an app that can handle all the different parts of the app that no one else is allowed to do

**Reply by erikn69** [reply]
There is many ways
[basic-usage/teams-permissions#content-defining-a-super-admin-on-teams](https://spatie.be/docs/laravel-permission/v6/basic-usage/teams-permissions#content-defining-a-super-admin-on-teams)

You could check "super admin" on main team, and use gate
[basic-usage/super-admin](https://spatie.be/docs/laravel-permission/v6/basic-usage/super-admin)

Choose the option that seems best to you, although you can achieve it only by putting a flag on the user who will be `super admin`

**Reply by sohailahmad07** [reply]
ok i have defined the super admin from basic usage in my app service provider but if like i said when assigning the super admin role it require the team_id but the super admin is not part of the team

is for defining super-admin on teams i believe what is happening there is

when you create a new team you just add the super-admin user to that team with roles 

like for example let's say i have a super-admin user when a new team is created  i add that super-admin to that team i think of that first but it wont work for me can my super admin can make another user super-admin so in that case what about them

**Reply by erikn69** [reply]
in that case maybe this
[basic-usage/teams-permissions#content-defining-a-super-admin-on-teams](https://spatie.be/docs/laravel-permission/v6/basic-usage/teams-permissions#content-defining-a-super-admin-on-teams)

**Reply by sohailahmad07** [reply]
@erikn69 thnx i appreciate you helping me i will explain the feature i am trying to implement and if you have any idea on how can I achieve that with spaite teams permission let me know

so by default teams are optional in my app Users don't have a team when the user register

so by default, you have these global roles like user, admin, and super-admin, and they have permission if they can do something or not in app

now there are some parts of the app that can not be accessed without a team so for that you have to join and team

once you join a team along with a default role you also get a team role which lets access the team stuff which will be up to the team leader how much access he wants to give to you for that team

i hope it helps

**Reply by erikn69** [reply]
In my opinion everything should work for you 👍

**Reply by sohailahmad07** [reply]
ok but its not so here my code in database seeder 

```
   User::factory()->create([
            'name' => 'Test User',
            'email' => 'test@example.com',
        ]);

        User::factory(10)->create();
        Role::create(['name' => 'Super Admin']);
        Role::create(['name' => 'Admin']);
        Role::create(['name' => 'User']);

//        setPermissionsTeamId(1);
        User::first()->assignRole('Super Admin');
        User::find(2)->assignRole('Admin');
        User::find(3)->assignRole('User');
```
Error I get
```
SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'organization_id' cannot be null (Connection: mariadb, SQL: insert into `model_has_roles` (`model_id`, `model_type`, `organization_id`, `role_id`) values (1, App\Models\User, ?, 1))
```


when i enable `setPermissionsTeamId(1)` it works but like is said those roles are not for the team what am i missing here

**Reply by erikn69** [reply]
Needs a default team
[basic-usage/teams-permissions#content-defining-a-super-admin-on-teams](https://spatie.be/docs/laravel-permission/v6/basic-usage/teams-permissions#content-defining-a-super-admin-on-teams)

You can't have them without at least one team, a main team

Maybe this package is not for your case, you could customize it, so it would work as you expect, or opt for another one that fits better.

**Reply by sohailahmad07** [reply]
well I don't know if its going to bite my later but i changed migrate after reading this https://github.com/spatie/laravel-permission/discussions/2446 discussion in which you can set a placeholder team id like 0 which does not exist and you can and since you can do that i changed the migration make `team_id` on `modal_has_roles` nullable and remove the teams' ids from a primary key and me some change to the roles method code is provided below

looking in tinker and everything seems to be working fine 

if you have a global role like super admin you don't need to set team_id to get the role you can just use 
$user->roles //  will give all of this users global roles
how every if you set team id and the use
setPermissionsTeamId(1)
$user->roles // it will give you both global roles and team base roles for that user 

which is kind of what I want

**I WILL UPDATE THIS IF I GET ANY SIDE EFFECTS IN MY PROJECT**





change i made in the permission migration
```
   if ($teams) {
                $table->unsignedBigInteger($columnNames['team_foreign_key'])->nullable(); // change here
                $table->index($columnNames['team_foreign_key'], 'model_has_roles_team_foreign_key_index');

                $table->primary([$pivotRole, $columnNames['model_morph_key'], 'model_type'], // change here
                    'model_has_roles_role_model_type_primary');
            } else {
                $table->primary([$pivotRole, $columnNames['model_morph_key'], 'model_type'],
                    'model_has_roles_role_model_type_primary');
            }
```
 
 In user Modal to overwrite the default roles behavior
 ```
     public function roles(): BelongsToMany
    {
        $relation = $this->morphToMany(
            config('permission.models.role'),
            'model',
            config('permission.table_names.model_has_roles'),
            config('permission.column_names.model_morph_key'),
            app(PermissionRegistrar::class)->pivotRole
        );
        if (!app(PermissionRegistrar::class)->teams) {
            return $relation;
        }

        $teamsKey = app(PermissionRegistrar::class)->teamsKey; // refactore

        $teamField = config('permission.table_names.roles') . '.' . $teamsKey;

      return $relation->wherePivot($teamsKey, getPermissionsTeamId())
            ->orWherePivotNull($teamsKey) // new condition
            ->wherePivot(config('permission.column_names.model_morph_key'), $this->id) // // new condition
            ->where(fn($q) => $q->whereNull($teamField)->orWhere($teamField, getPermissionsTeamId()));
    }
 ```
